query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+            self.extracted_types.insert(var.clone(), alias.for_type_tag());
+        }
+        self.bindings.entry(var).or_insert(vec![]).push(alias);
+    }
+
+    pub fn constrain_column_to_entity(&mut self, table: TableAlias, column: DatomsColumn, entity: i64) {
+        self.wheres.push(ColumnConstraint::EqualsEntity(QualifiedAlias(table, column), entity))
+    }
+
+    pub fn constrain_attribute(&mut self, table: TableAlias, attribute: i64) {
+        self.constrain_column_to_entity(table, DatomsColumn::Attribute, attribute)
+    }
+
+    pub fn constrain_column_to_constant(&mut self, table: TableAlias, column: DatomsColumn, constant: TypedValue) {
+        self.wheres.push(ColumnConstraint::EqualsValue(QualifiedAlias(table, column), constant))
+    }
+
+    /// Returns false if it's impossible for this type to apply.
+    /// Constraints the var if there's no existing type.
+    fn constrain_var_to_type(&mut self, variable: Variable, this_type: ValueType) -> bool {
+        // Is there an existing binding for this variable?
+        let types_entry = self.known_types.entry(variable);
+        match types_entry {
+            // If so, the types must match.
+            Entry::Occupied(entry) =>
+                *entry.get() == this_type,
+            // If not, record the one we just determined.
+            Entry::Vacant(entry) => {
+                entry.insert(this_type);
+                true
+            }
+        }
+    }
+
+    /// Ensure that the given place has the correct types to be a tx-id.
+    /// Right now this is mostly unimplemented: we fail hard if anything but a placeholder is
+    /// present.
+    fn mark_as_tx(&mut self, tx: &PatternNonValuePlace) {
+        match *tx {
+            PatternNonValuePlace::Placeholder => (),
+            _ => unimplemented!(),           // TODO
+        }
+    }
+
+    /// Ensure that the given place can be an entity, and is congruent with existing types.
+    /// This is used for `entity` and `attribute` places in a pattern.
+    fn mark_as_ref(&mut self, value: &PatternNonValuePlace) {
+        // If it's a variable, record that it has the right type.
+        // Ident or attribute resolution errors (the only other check we need to do) will be done
+        // by the caller.
+        if let &PatternNonValuePlace::Variable(ref v) = value {
+            if !self.constrain_var_to_type(v.clone(), ValueType::Ref) {
+                self.mark_known_empty("Couldn't constrain var to Ref.");
+            }
+        }
+    }
+
+    fn mark_known_empty(&mut self, why: &str) {
+        self.is_known_empty = true;
+        println!("{}", why);                   // TODO: proper logging.
+    }
+
+    fn entid_for_ident<'s, 'a>(&self, schema: &'s Schema, ident: &'a NamespacedKeyword) -> Option<i64> {
+        schema.get_entid(&ident.to_string())      // TODO: #291.
+    }
+
+    fn attribute_for_ident<'s, 'a>(&self, schema: &'s Schema, ident: &'a NamespacedKeyword) -> Option<&'s Attribute> {
```

Comment:
This should be in `Schema`, I think.  I intend to pass `&Attribute` refs around in the DB layer eventually too.